packages/block-editor/src/components/provider/index.js

```diff
@@ -120,10 +124,13 @@ class BlockEditorProvider extends Component {
 				blocks = newBlocks;
 				isPersistent = newIsPersistent;
 
+				const selectionStart = getSelectionStart();
+				const selectionEnd = getSelectionEnd();
+
 				if ( isPersistent ) {
-					onChange( blocks );
+					onChange( blocks, selectionStart, selectionEnd );
```

Comment:
I see the motivation to avoid two separate state changes, but I am very reluctant to start introducing additional arguments here which are relevant for specific use-cases like managing selection, vs. discrete actions which manage those states. Otherwise the interface for `resetEditorBlocks` becomes more difficult to understand and to use, and we set a precedent where in the future we feel justified to add even more arguments here. Already recently there was some similar proposal in the now-closed #16075 to add an options argument to specify the cause of the change.

At the very worst, I would suggest that we anticipate needing more options here, and form it as an object.

At best, I wonder if we find another way to address the problem we're trying to resolve with lumping the two discrete actions into one; maybe we can explore an implementation of "action transactions", where we can apply multiple actions as at-most a single state change.

Example:

```js
wp.data.transact( () => {
	wp.data.dispatch( 'core/editor' ).resetEditorBlocks( blocks );
	wp.data.dispatch( 'core/editor' ).setSelection( selection );
} );
```

Reply:
Hmm, outside some difficulty in implementing this transactions behavior, I wonder if it does more harm than good by obfuscating actions which occur. I wonder: What is the main concern we have with these being two separate actions in the first place?

Reply:
@aduth Initially I did it exactly like that. The only issue is that the history reducer then needs to ignore any selection changes and that order matters. So you have to set selection first, then set the content. That's fine I guess.